src/treelearner/col_sampler.hpp

```diff
@@ -83,32 +88,92 @@ class ColSampler {
     }
   }
 
-  std::vector<int8_t> GetByNode() {
-    if (fraction_bynode_ >= 1.0f) {
-      return std::vector<int8_t>(train_data_->num_features(), 1);
+  std::vector<int8_t> GetByNode(const Tree* tree, int leaf) {
+    // get interaction constraints for current branch
+    std::unordered_set<int> allowed_features;
```

Comment:
is that possible to cache the used features of all branches, rather than re-generate them every time?

Reply:
I've changed the code so that we keep track of the features in each branch as we build the tree.

Reply:
Thanks so much! However, it seems the tracking is still enabled even when interaction constraint is not used?

Reply:
Thanks, I changed it now so it only keeps track if interaction constraint is enabled.